test/functional/cursor.test.js

```diff
@@ -3901,6 +3901,43 @@ describe('Cursor', function () {
     });
   });
 
+  // NODE-2035
+  it('should propagate error when exceptions are thrown from an awaited forEach call', function () {
+    const configuration = this.configuration;
+    const client = configuration.newClient({ w: 1 }, { maxPoolSize: 1 });
+    const docs = [{ a: 1 }, { a: 2 }, { a: 3 }];
+    let collection;
+    let db;
+    return client
+      .connect()
+      .then(() => {
+        db = client.db(configuration.db);
+        collection = db.collection('cursor_session_tests2');
+        return collection.insertMany(docs);
+      }, console.error)
+      .then(() => {
+        const cursor = collection.find({});
+        async function testAsync() {
+          let promiseResult;
+          try {
+            promiseResult = await cursor
+              .forEach(() => {
+                throw new Error('FAILURE IN FOREACH CALL');
+              })
+              .catch(err => {
+                expect(err.message).to.deep.equal('FAILURE IN FOREACH CALL');
+              });
+          } catch (err) {
+            expect(err).to.be.undefined;
+          }
```

Comment:
```suggestion
        const cursor = collection.find({});
        expect(() => {
          cursor
              .forEach(() => {
                throw new Error('FAILURE IN FOREACH CALL');
              })
              .catch(err => {
                expect(err.message).to.deep.equal('FAILURE IN FOREACH CALL');
              });
        }).to.not.throw();
```

We can reduce some of the logic here, if we pass a function to chai with expect(...).to.not.throw(); then we can confirm that the forEach doesn't throw synchronously and still check that the error is correctly propagated to the catch callback. (ie. its part of the promise chain).

Reply:
Fwiw, it might also be nice here to check that the `catch` handler is actually called?